src/System.Xml.XDocument/src/System/Xml/Linq/Extensions.cs

```diff
@@ -269,7 +268,7 @@ public static IEnumerable<XElement> DescendantsAndSelf(this IEnumerable<XElement
         /// </returns>
         public static IEnumerable<T> InDocumentOrder<T>(this IEnumerable<T> source) where T : XNode
         {
-            return Enumerable.OrderBy(source, n => (XNode)n, XNode.DocumentOrderComparer);
+            return new OrderedNodeEnumerable<T>(source);
```

Comment:
Instead of a separate `OrderedNodeEnumerable` class, why not something like the following?

```c#
    public static IEnumerable<T> InDocumentOrder<T>(this IEnumerable<T> source) where T : XNode
    {
		if (source == null) throw new ArgumentNullException("source");
        return InDocumentOrderIterator(source);
    }
    
    private static IEnumerable<T> InDocumentOrderIterator<T>(this IEnumerable<T> source) where T : XNode
    {
        ... iterator implementation here ...
    }
```

The benefit is that it avoids the extra enumerator allocation when `GetEnumerator` is called on the same thread (common case), as the iterator generated by the compiler will return the same object.